src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenNullCheckedParameterTests.cs

```diff
@@ -1456,14 +1547,189 @@ class C
+}");
+        }
+
+        [Fact]
+        public void TestNullableValueType()
+        {
+            var source = @"
+class C
+{
+    public void M(int? i!!) // 1
+    {
+    }
+}";
+            var verifier = CompileAndVerify(source);
+            verifier.VerifyDiagnostics(
+                // (4,24): warning CS8995: Nullable value type 'int?' is null-checked and will throw if null.
+                //     public void M(int? i!!) // 1
+                Diagnostic(ErrorCode.WRN_NullCheckingOnNullableValueType, "i").WithArguments("int?").WithLocation(4, 24));
+            verifier.VerifyIL("C.M(int?)", @"
+{
+  // Code size       21 (0x15)
+  .maxstack  1
+  IL_0000:  ldarga.s   V_1
+  IL_0002:  call       ""bool int?.HasValue.get""
+  IL_0007:  brtrue.s   IL_0014
+  IL_0009:  ldstr      ""i""
+  IL_000e:  newobj     ""System.ArgumentNullException..ctor(string)""
+  IL_0013:  throw
+  IL_0014:  ret
+}");
+        }
+
+        [Fact]
+        public void NullCheckedPointer()
+        {
+            var source = @"
+using System;
+
+
+class C
+{
+    static unsafe void M1(void* ptr!!) { }
+    static unsafe void M2(int* ptr!!) { }
+    static unsafe void M3(delegate*<int, void> funcPtr!!) { }
+
+    public static void M0(int x) { }
+
+    public static unsafe void Main()
+    {
+        int x = 1;
+        try
+        {
+            M1(&x);
+            Console.Write(1);
+            M1(null);
+        }
+        catch
+        {
+            Console.Write(2);
+        }
+
+        try
+        {
+            M2(&x);
+            Console.Write(3);
+            M2(null);
+        }
+        catch
+        {
+            Console.Write(4);
+        }
+
+        try
+        {
+            M3(&M0);
+            Console.Write(5);
+            M3(null);
+        }
+        catch
+        {
+            Console.Write(6);
+        }
+    }
+}";
+            var verifier = CompileAndVerify(source, options: TestOptions.UnsafeDebugExe, verify: Verification.Skipped, expectedOutput: "123456");
+            verifier.VerifyDiagnostics();
+            verifier.VerifyIL("C.M1", @"
+{
+  // Code size       16 (0x10)
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  brtrue.s   IL_000e
+  IL_0003:  ldstr      ""ptr""
+  IL_0008:  newobj     ""System.ArgumentNullException..ctor(string)""
+  IL_000d:  throw
+  IL_000e:  nop
+  IL_000f:  ret
+}");
+            verifier.VerifyIL("C.M2", @"
+{
+  // Code size       16 (0x10)
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  brtrue.s   IL_000e
+  IL_0003:  ldstr      ""ptr""
+  IL_0008:  newobj     ""System.ArgumentNullException..ctor(string)""
+  IL_000d:  throw
+  IL_000e:  nop
+  IL_000f:  ret
+}");
+            verifier.VerifyIL("C.M3", @"
+{
+  // Code size       16 (0x10)
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  brtrue.s   IL_000e
+  IL_0003:  ldstr      ""funcPtr""
+  IL_0008:  newobj     ""System.ArgumentNullException..ctor(string)""
+  IL_000d:  throw
+  IL_000e:  nop
+  IL_000f:  ret
 }");
         }
```

Comment:
>} [](http://example.com/codeflow?start=8&length=1)

Consider verifying that we did not generate `ThrowIfNull()` and `Throw()` methods in these last three tests that inline the checks.
```csharp
verifier.VerifyMissing("<PrivateImplementationDetails>.ThrowIfNull");
verifier.VerifyMissing("<PrivateImplementationDetails>.Throw");
```

Reply:
Done